### PR TITLE
Utilities: Port tar to LibMain

### PR DESCRIPTION
**Pull request by cocateh**
Ported tar to LibMain and changed it to use Core::System syscalls as
well as fixed Utilities/CMakefile.txt unalphabetical position of one of
the entries.

### CODE REVIEW

File: Userland/Utilities/tar.cpp [hunk 1]
```diff
@@ -104,39 +97,28 @@ int main(int argc, char** argv)
...
                 case Archive::TarFileType::Directory: {
                     Core::File::ensure_parent_directories(absolute_path);
 
-                    if (mkdir(absolute_path.characters(), header.mode()) && errno != EEXIST) {
-                        perror("mkdir");
-                        return 1;
+                    auto result_or_error = Core::System::mkdir(absolute_path.characters(), header.mode());
+                    if (result_or_error.is_error() && result_or_error.error().code() != EEXIST) {
+                        auto error = result_or_error.error();
+                        warnln("mkdir: {}", error);
```
Comment (LucasChollet) [inline Userland/Utilities/tar.cpp:121]: You are changing the semantic here.
You should probably return the error.

Comment (cocateh) [reply Userland/Utilities/tar.cpp:121]: Resolved, thanks!

File: Userland/Utilities/CMakeLists.txt [hunk 2]
```diff
@@ -149,6 +149,7 @@ target_link_libraries(open LibDesktop LibMain)
 target_link_libraries(pape LibGUI LibMain)
 target_link_libraries(passwd LibCrypt LibMain)
 target_link_libraries(paste LibGUI LibCore LibMain)
+target_link_libraries(pathchk LibMain)
```
Comment (kleinesfilmroellchen) [inline Userland/Utilities/CMakeLists.txt:152]: If you're linting CMake files, please put that into another commit.

File: Userland/Utilities/tar.cpp [hunk 3]
```diff
@@ -104,39 +97,29 @@ int main(int argc, char** argv)
...
                 }
                 case Archive::TarFileType::SymLink: {
                     Core::File::ensure_parent_directories(absolute_path);
 
-                    if (symlink(header.link_name().to_string().characters(), absolute_path.characters())) {
-                        perror("symlink");
-                        return 1;
-                    }
-
+                    TRY(Core::System::symlink(header.link_name().to_string().characters(), absolute_path.characters()));
```
Comment (kleinesfilmroellchen) [inline Userland/Utilities/tar.cpp:112]: Why should a Core:System API require char*?

File: Userland/Utilities/tar.cpp (hunk 1 above)
Comment (LucasChollet) [reply Userland/Utilities/tar.cpp:121]: The error will probably be displayed twice, you should let the caller manage the error.
```cpp
if (result_or_error.is_error() && result_or_error.error().code() != EEXIST)
    return result_or_error.error();
```

File: Userland/Utilities/tar.cpp [hunk 4]
```diff
@@ -197,23 +170,27 @@ int main(int argc, char** argv)
                 return;
             }
 
-            struct stat statbuf;
-            if (lstat(path.characters(), &statbuf) < 0) {
+            auto statbuf_or_error = Core::System::lstat(path.characters());
+            if (statbuf_or_error.is_error()) {
                 warnln("Failed stating {}", path);
                 return;
```
Comment (kleinesfilmroellchen) [inline Userland/Utilities/tar.cpp:176]: This should return the error as well.

File: Userland/Utilities/tar.cpp [hunk 5]
```diff
@@ -197,23 +170,27 @@ int main(int argc, char** argv)
...
             }
 
-            struct stat statbuf;
-            if (lstat(path.characters(), &statbuf) < 0) {
+            auto statbuf_or_error = Core::System::lstat(path.characters());
```
Comment (kleinesfilmroellchen) [inline Userland/Utilities/tar.cpp:173]: Again, why does this require a char*?

File: Userland/Utilities/tar.cpp [hunk 6]
```diff
@@ -197,23 +170,27 @@ int main(int argc, char** argv)
...
                 outln("{}", canonicalized_path);
         };
 
         auto add_directory = [&](String path, auto handle_directory) -> void {
-            struct stat statbuf;
-            if (lstat(path.characters(), &statbuf) < 0) {
+            auto statbuf_or_error = Core::System::lstat(path.characters());
+            if (statbuf_or_error.is_error()) {
                 warnln("Failed stating {}", path);
                 return;
```
Comment (kleinesfilmroellchen) [inline Userland/Utilities/tar.cpp:190]: See above

File: Userland/Utilities/tar.cpp [hunk 7]
```diff
@@ -239,7 +216,6 @@ int main(int argc, char** argv)
         }
 
         tar_stream.finish();
-
```
Comment (kleinesfilmroellchen) [inline Userland/Utilities/tar.cpp:242]: Please keep this line.

File: Userland/Utilities/tar.cpp (hunk 1 above)
Comment (cocateh) [reply Userland/Utilities/tar.cpp]: fixed

File: Userland/Utilities/CMakeLists.txt (hunk 2 above)
Comment (cocateh) [reply Userland/Utilities/CMakeLists.txt:152]: Fixed.

File: Userland/Utilities/tar.cpp (hunk 7 above)
Comment (cocateh) [reply Userland/Utilities/tar.cpp]: Fixed.

File: Userland/Utilities/tar.cpp (hunk 6 above)
Comment (cocateh) [reply Userland/Utilities/tar.cpp]: Fixed.

File: Userland/Utilities/tar.cpp (hunk 4 above)
Comment (cocateh) [reply Userland/Utilities/tar.cpp]: Fixed.

File: Userland/Utilities/tar.cpp (hunk 3 above)
Comment (cocateh) [reply Userland/Utilities/tar.cpp]: Fixed.

File: Userland/Utilities/tar.cpp [hunk 8]
```diff
@@ -223,18 +195,19 @@ int main(int argc, char** argv)
...
                 auto child_path = it.next_full_path();
                 if (!Core::File::is_directory(child_path)) {
-                    add_file(child_path);
+                    TRY(add_file(child_path));
                 } else {
-                    handle_directory(child_path, handle_directory);
+                    TRY(handle_directory(child_path, handle_directory));
                 }
             }
+            return ErrorOr<void>();
```
Comment (kleinesfilmroellchen) [inline Userland/Utilities/tar.cpp:203]: Please use the implicit default constructor through `{}`

**Review by kleinesfilmroellchen — COMMENTED** [review]
Just one small thing

File: Userland/Utilities/tar.cpp (hunk 8 above)
Comment (cocateh) [reply Userland/Utilities/tar.cpp]: Done.

**Review by kleinesfilmroellchen — APPROVED** [review]
